### ISSUE TITLE
Multichannel: cudacpp implementation fails for diagrams producing more than one amplitude

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#472]
Hi @oliviermattelaer this is a followup to #419 and specifically https://github.com/madgraph5/madgraph4gpu/issues/419#issuecomment-1145076040

Essentially, in gg->ttgg I find that, even with SD=1 and LIMHEL=0, I still get a large difference between fortran and cudacpp MEs. After looking at the code, I would say that the reason is quite obvious: the cudacpp implementation does not handle correctly the diagrams that produce more than one amplitude. Actually I am not sure what this means: I guess that the same Feynman diagram may actually need different VVV cals depending on the specific color combination? But then how should the numerators and denominators of multichannel be handled?

Anyway: I think this is a problem in the upstream rev1009 for standalone_gpu, not something that I introdcued in the standalone_cudacpp derived from that. This is what happens in ggttgg.madgpu from standalone_gpu:

```
  VVV1P0_1(w[0], w[1], cxtype(cIPC[0], cIPC[1]), 0., 0., w[6]); 
  FFV1P0_3(w[3], w[2], cxtype(cIPC[2], cIPC[3]), 0., 0., w[7]); 
  // Amplitude(s) for diagram number 1
  VVVV1_0(w[6], w[7], w[4], w[5], cxtype(cIPC[4], cIPC[5]), &amp[0]); 
  jamp[0] += +cxtype(0, 1) * amp[0]; 
  jamp[1] += -cxtype(0, 1) * amp[0]; 
  jamp[6] += -cxtype(0, 1) * amp[0]; 
  jamp[7] += +cxtype(0, 1) * amp[0]; 
  jamp[16] += -cxtype(0, 1) * amp[0]; 
  jamp[17] += +cxtype(0, 1) * amp[0]; 
  jamp[22] += +cxtype(0, 1) * amp[0]; 
  jamp[23] += -cxtype(0, 1) * amp[0]; 
  VVVV3_0(w[6], w[7], w[4], w[5], cxtype(cIPC[4], cIPC[5]), &amp[0]); 
  jamp[0] += +cxtype(0, 1) * amp[0]; 
  jamp[6] += -cxtype(0, 1) * amp[0]; 
  jamp[12] += -cxtype(0, 1) * amp[0]; 
  jamp[14] += +cxtype(0, 1) * amp[0]; 
  jamp[18] += -cxtype(0, 1) * amp[0]; 
  jamp[20] += +cxtype(0, 1) * amp[0]; 
  jamp[22] += +cxtype(0, 1) * amp[0]; 
  jamp[23] += -cxtype(0, 1) * amp[0]; 
  VVVV4_0(w[6], w[7], w[4], w[5], cxtype(cIPC[4], cIPC[5]), &amp[0]); 
  jamp[1] += +cxtype(0, 1) * amp[0]; 
  jamp[7] += -cxtype(0, 1) * amp[0]; 
  jamp[12] += -cxtype(0, 1) * amp[0]; 
  jamp[14] += +cxtype(0, 1) * amp[0]; 
  jamp[16] += +cxtype(0, 1) * amp[0]; 
  jamp[17] += -cxtype(0, 1) * amp[0]; 
  jamp[18] += -cxtype(0, 1) * amp[0]; 
  jamp[20] += +cxtype(0, 1) * amp[0]; 
  VVV1P0_1(w[6], w[4], cxtype(cIPC[0], cIPC[1]), 0., 0., w[8]); 
  // Amplitude(s) for diagram number 2
  VVV1_0(w[7], w[5], w[8], cxtype(cIPC[0], cIPC[1]), &amp[0]); 
  if(channel_id == 1)
  {
    multi_chanel_num += conj(amp[0]) * amp[0]; 
  }; 
  multi_chanel_denom += conj(amp[0]) * amp[0]; 
  jamp[0] += +cxtype(0, 1) * amp[0]; 
  jamp[6] += -cxtype(0, 1) * amp[0]; 
  jamp[12] += -cxtype(0, 1) * amp[0]; 
  jamp[14] += +cxtype(0, 1) * amp[0]; 
  jamp[18] += -cxtype(0, 1) * amp[0]; 
  jamp[20] += +cxtype(0, 1) * amp[0]; 
  jamp[22] += +cxtype(0, 1) * amp[0]; 
  jamp[23] += -cxtype(0, 1) * amp[0]; 
  VVV1P0_1(w[6], w[5], cxtype(cIPC[0], cIPC[1]), 0., 0., w[9]); 
  // Amplitude(s) for diagram number 3
  VVV1_0(w[7], w[4], w[9], cxtype(cIPC[0], cIPC[1]), &amp[0]); 
  if(channel_id == 2)
  {
    multi_chanel_num += conj(amp[0]) * amp[0]; 
  }; 
  multi_chanel_denom += conj(amp[0]) * amp[0]; 
  jamp[1] += +cxtype(0, 1) * amp[0]; 
  jamp[7] += -cxtype(0, 1) * amp[0]; 
  jamp[12] += -cxtype(0, 1) * amp[0]; 
  jamp[14] += +cxtype(0, 1) * amp[0]; 
  jamp[16] += +cxtype(0, 1) * amp[0]; 
  jamp[17] += -cxtype(0, 1) * amp[0]; 
  jamp[18] += -cxtype(0, 1) * amp[0]; 
  jamp[20] += +cxtype(0, 1) * amp[0]; 
...
  // Amplitude(s) for diagram number 113
  FFV1_0(w[12], w[14], w[1], cxtype(cIPC[2], cIPC[3]), &amp[0]); 
  if(channel_id == 105)
  {
    multi_chanel_num += conj(amp[0]) * amp[0]; 
  }; 
  multi_chanel_denom += conj(amp[0]) * amp[0]; 
...
```

Essentially, the algorithm presently calculates the numerator and denominator ONLY from those diagrams that have a single amplitude connected to it. Those diagrams with multiple amplitudes have no contributions to numerator and denominator. Incidentally, this means also that the indexing of channels is wrong, for instance here 'channel 1' actually corresponds to 'diagram 2', which is the first diagram with only one amplitude associated, as 'diagram 1' is skipped (see for instance how channel 105 comes to correspond to diagram 113).

Can you suggest a way to fix this please? I am not entirely sure how to do this myself because I am not sure how to handle the three colors for each diagram: should the numerator and the denominator include a contribution that is the sum of the squares of the three colors? If this is teh case I can try to patch it....

Thanks!
Andrea

**Comment by valassi** [comment]
> Can you suggest a way to fix this please? I am not entirely sure how to do this myself because I am not sure how to handle the three colors for each diagram: should the numerator and the denominator include a contribution that is the sum of the squares of the three colors? If this is teh case I can try to patch it....

In other words, is it enough to add the six lines below (three colors times one numerator and one denominator)?
```
      // Amplitude(s) for diagram number 1
      VVVV1_0<W_ACCESS, A_ACCESS, CD_ACCESS>( w_fp[6], w_fp[7], w_fp[4], w_fp[5], COUPs[2], &amp_fp[0] );
      if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
      jamp_sv[0] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[1] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[6] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[7] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[16] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[17] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[22] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[23] -= cxtype( 0, 1 ) * amp_sv[0];
      VVVV3_0<W_ACCESS, A_ACCESS, CD_ACCESS>( w_fp[6], w_fp[7], w_fp[4], w_fp[5], COUPs[2], &amp_fp[0] );
      if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
      jamp_sv[0] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[6] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[12] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[14] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[18] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[20] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[22] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[23] -= cxtype( 0, 1 ) * amp_sv[0];
      VVVV4_0<W_ACCESS, A_ACCESS, CD_ACCESS>( w_fp[6], w_fp[7], w_fp[4], w_fp[5], COUPs[2], &amp_fp[0] );
      if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
      jamp_sv[1] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[7] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[12] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[14] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[16] += cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[17] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[18] -= cxtype( 0, 1 ) * amp_sv[0];
      jamp_sv[20] += cxtype( 0, 1 ) * amp_sv[0];
```

**Comment by valassi** [comment]
Trying to understand what happens in fortran
```
      IF (MULTI_CHANNEL) THEN
        XTOT=0D0
        DO I=1,LMAXCONFIGS
          J = CONFSUB(1, I)
          IF (J.NE.0) THEN
            IF(SDE_STRAT.EQ.1) THEN
              AMP2(J) = AMP2(J) * GET_CHANNEL_CUT(P, I)
              XTOT=XTOT+AMP2(J)
            ELSE
              AMP2(J) = GET_CHANNEL_CUT(P, I)
              XTOT=XTOT+AMP2(J)
            ENDIF
          ENDIF
        ENDDO
        IF (XTOT.NE.0D0) THEN
          ANS=ANS*AMP2(CHANNEL)/XTOT
        ELSE IF(ANS.NE.0D0) THEN
          IF(NB_FAIL.GE.10)THEN
            WRITE(*,*) 'Problem in the multi-channeling. All amp2 are'
     $       //' zero but not the total matrix-element'

            STOP 1
          ELSE
            NB_FAIL = NB_FAIL +1
          ENDIF
        ENDIF
      ENDIF
```

Essentially the numerator for channel i is AMP2(i) where AMP2 is one of the outputs of MATRIX1. Then the denominator is the sum of the AMP2(i) for all i's.

**Comment by valassi** [comment]
But I am totally confused. For channel 1 in gg to ggtt, I do not seem to find anywhere in the Fortran where AMP2(1) is defined?...

**Comment by valassi** [comment]
Does this simply mean that channel=1 is NOT a valid choice for gg to ttgg?...

**Comment by valassi** [comment]
Trying to make sense of it, for ggttgg

In maxconfigs.inc
```
      INTEGER LMAXCONFIGS
      PARAMETER(LMAXCONFIGS=201)
```

In maxamps.inc
```
      INTEGER    MAXAMPS, MAXFLOW, MAXPROC, MAXSPROC
      PARAMETER (MAXAMPS=123, MAXFLOW=24)
      PARAMETER (MAXPROC=1, MAXSPROC=1)
```

In matrix1.f
```
      DOUBLE PRECISION AMP2(MAXAMPS), JAMP2(0:MAXFLOW)
...
      INTEGER    NGRAPHS
      PARAMETER (NGRAPHS=159)
      INTEGER    NDIAGS
      PARAMETER (NDIAGS=123)
...
      IF (MULTI_CHANNEL) THEN
        DO I=1,NDIAGS
          AMP2(I)=0D0
        ENDDO
...
      IF (MULTI_CHANNEL) THEN
        XTOT=0D0
        DO I=1,LMAXCONFIGS
          J = CONFSUB(1, I)
          IF (J.NE.0) THEN
            IF(SDE_STRAT.EQ.1) THEN
              AMP2(J) = AMP2(J) * GET_CHANNEL_CUT(P, I)
              XTOT=XTOT+AMP2(J)
            ELSE
              AMP2(J) = GET_CHANNEL_CUT(P, I)
              XTOT=XTOT+AMP2(J)
            ENDIF
          ENDIF
        ENDDO
        IF (XTOT.NE.0D0) THEN
          ANS=ANS*AMP2(CHANNEL)/XTOT
```

In CPPProcess.cc
```
      // *** DIAGRAM 2 OF 123 ***

      // Wavefunction(s) for diagram number 2
      VVV1P0_1<W_ACCESS, CD_ACCESS>( w_fp[6], w_fp[4], COUPs[0], 0., 0., w_fp[8] );

      // Amplitude(s) for diagram number 2
      VVV1_0<W_ACCESS, A_ACCESS, CD_ACCESS>( w_fp[7], w_fp[5], w_fp[8], COUPs[0], &amp_fp[0] );
      if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
...
      // *** DIAGRAM 113 OF 123 ***

      // Wavefunction(s) for diagram number 113
      // (none)

      // Amplitude(s) for diagram number 113
      FFV1_0<W_ACCESS, A_ACCESS, CD_ACCESS>( w_fp[12], w_fp[14], w_fp[1], COUPs[1], &amp_fp[0] );
      if( channelId == 105 ) numerators_sv += cxabs2( amp_sv[0] );
      if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
```

I would tend to think that:
- there are 123 diagrams
- AMP2 is dimensioned for 123 diagrams, which are potential "channels"
- the fact that two separate variables NDIAGS and MAXAMPS are both equal to 123 and are used more or less equivalently is very confusing (AMP2 is dimensioned from MAXAMPS, but then the loop over elements to zero them goes through NDIAGS)
- actually, the number of channels is not 123! there are only 105 'channels' (at least judging from CPPProcess)

Then it gets more confusing, and here is probably the bug
- when I enter "1" in the input file as channel to enhance, then when I print out channel from fortran I get "2"?
- so maybe somehow 1 is translated to the first valid diagram that is a channel, ie 2?
- in cpp instead the first valid diagram that is a channel, diagram 2, is assigned channelId=1

Todo
- what happens if I enter a channel that is too large in the input file, eg 123?
- who changes the input 1 to 2 in the fortran
- would it be enough to replace channelId=1 by channelId=2 for diagram 2 in CppProcess,cc?

**Comment by valassi** [comment]
Ok my working hypothesis above seems correct
- while channelId is in the range 1 to 123, only some channelId are valid
- further hypothesis: the channelId actually IS the diagram number (and digarams like diagram 1 which have multiple amplitudes simply have no associated channel), in this case it will probably be enough to change channelId=1 to the diagram number on CppProcess
- the input channelId in the input file is NOT The diagram number, eg 1 is shifted to channelId=2 (I have not understood who does it, but it's not really relevant); in particular, using 123 as input fails, because there are only 105 channels, even if there are 123 diagrams

https://github.com/madgraph5/madgraph4gpu/pull/465/commits/7765c362fa8ff89fe44934d0e07248a52194a166
```
(1) using this input

4 ! Number of events in a single C++ iteration (nb_page_loop)
4 1 1 ! Number of events and max and min iterations
0.000001 ! Accuracy (ignored because max iterations = min iterations)
0 ! Grid Adjustment 0=none, 2=adjust (NB if = 0, ftn26 will still be used if present)
1 ! Suppress Amplitude 1=yes (i.e. use MadEvent single-diagram enhancement)
0 ! Helicity Sum/event 0=exact
1 ! Channel number (1-N) for single-diagram enhancement multi-channel (NB NOT IGNORED even if suppress amplitude is 0!)

gives

channelId=2, numerators_sv=1.50103e-09
channelId=2, numerators_sv=7.8872e-12
WARNING! ( 1/20) Deviation more than 5E-5   1  0.31473953E-11  0.25055756E-11          0.79607908844
WARNING! ( 2/20) Deviation more than 5E-5   2  0.86377709E-14  0.59380144E-14          0.68744754341
WARNING! ( 3/20) Deviation more than 5E-5   3  0.56728598E-11  0.28663323E-10          5.05271134415
WARNING! ( 4/20) Deviation more than 5E-5   4  0.67945797E-10  0.36986147E-12          0.00544347826
 MULTI_CHANNEL = TRUE
 CHANNEL_ID =           2
 channel amp2(channel) ans           2   3.7265763466774809E-011   1.8832004853935949E-013
 channel amp2(channel) ans           2   5.7894820068594045E-010   4.1150126202644131E-011
 channel amp2(channel) ans           2   3.3605847937929455E-010   7.2785908222583483E-012
 channel amp2(channel) ans           2   3.2919737045219806E-010   1.1434590560834042E-011
channelId=2, numerators_sv=3.5247e-11
channelId=2, numerators_sv=2.16182e-13
channelId=2, numerators_sv=9.5787e-13
channelId=2, numerators_sv=6.559e-12
channelId=2, numerators_sv=1.04863e-10
channelId=2, numerators_sv=4.32568e-13

(2) changing the input to have channel=123 essentially fails

 No points passed cuts!
 Loosen cuts or increase max_events     5000000
 Deleting file events.lhe
Note: The following floating-point exceptions are signalling: IEEE_UNDERFLOW_FLAG IEEE_DENORMAL
```

**Comment by valassi** [comment]
This is now fixed by PR #465, specifically by https://github.com/valassi/madgraph4gpu/commit/384b7822177aea8ec669d3e28351481e318c0b9f

In cudacpp I am now using as channelId the diagram number. In my understanding and as discussed with Olivier, this seems to be the case also in Fortran.

Note that 
- indeed there can be more than one amplitude per diagram
- indeed all channels are diagrams, but there are some diagrams that are not channels
- it is not 100% clear whether a diagram is not a channel IF AND ONLY IF it has more than one amplitude: there may be niche cases (eg CKM according to Olivier) where one diagram has more than one amplitude and still represents a channel... in that case the issues I described at the top, i.e. how to build numerator and denominator from more than one amplitudes, do apply (sum the squares of the amplitudes? sume the amplitudes an dthen square? do something more complex?)... this possible issue is not handled here, it will need to be handled if and when this arises for some specific processes, but it does not happen for the processes I am considering here
